### ISSUE TITLE
String interpolation

### ISSUE THREAD

**Issue by borkdude** [clavascript/clavascript#32]
@corasaurus-hex String interpolation is a different issue: this issue is just about generating more efficient code for `(str foo bar)` => `cherry/str(foo,bar)` vs `foo + bar`.
I think string interpolation should also be a built-in feature that compiles directly to JS string interpolation. We can support this using a reader template `#i "foo ${name}"`. I'll make a different issue for that.

_Originally posted by @borkdude in https://github.com/clavascript/clavascript/issues/10#issuecomment-1217531806_

**Comment by lilactown** [comment]
Nitpicking: `#i` is a bit ambiguous, what about `#str`?

**Comment by borkdude** [comment]
Yes, `#i` isn't a good choice. Not sure about `#str`, maybe too general?. What about `#<<`? `<<` was the string interpolation macro in clojure.contrib. Too cryptic? 

```
#<< "hello ${name}"
```

**Comment by borkdude** [comment]
Or `#$ "hello ${name}`: too Perly...

**Comment by lilactown** [comment]
Outside of the reader tag we use, how ought we to handle custom tag functions? Something that would expand to

```javascript
html`<div style="color: red;">hello ${user}!</div>`
```

**Comment by lilactown** [comment]
perhaps `#tmpl (html "<div style="color: red;">hello ${user}!</div>")`

**Comment by corasaurus-hex** [comment]
maybe `#\`html` ? Some way to indicate that these are backticks?

**Comment by corasaurus-hex** [comment]
oh dang that's not bad @lilactown, I like it

**Comment by corasaurus-hex** [comment]
how does that generalize to bare template strings, though?

**Comment by lilactown** [comment]
i'm a bit ignorant of how the reader works is but i think we could have it support both bare strings as well as lists with a tag function

```clojure
#tmpl "Hello, ${user}!"

#tmpl (html "<div style="color: red;">hello ${user}!</div>")
```

I'm using `tmpl` as a stand in for whatever we decide to use.

**Comment by lilactown** [comment]
Thinking out loud, I'm worried about how we'll handle complex cases than just interpolating a variable. For instance, how should we handle expressions?

```clojure
#tmpl "Hello, ${(:name user)}!"
```

Perhaps we should lean into the tag accepting a collection, and instead of writing the `${}' syntax directly we write lists

```clojure
#tmpl ("Hello, " user-name "!")
#tmpl ("Hello, " (:name user) "!")
#tmpl (html "<div style=\"color: red;\">Hello, " user-name "!</div>")
```

**Comment by borkdude** [comment]
> #tmpl ("Hello, " user-name "!")

Hmm, this isn't any better to me than just using `str`. The point of `"${foo}"` is that it's easier to write.

We could support `#tmpl "Hello ${name}"` and `#tmpl [:html "<pre>${name}</pre>]`.

**Comment by lilactown** [comment]
@borkdude how should we handle arbitrary expressions inside of the template string, like I was talking about above?
